_posts/2023-03-27-group-equivariant-cnn-2/group-equivariant-cnn-2.Rmd

```diff
@@ -0,0 +1,374 @@
+
+Here is a goat.
+
+```{r}
+img_path <- system.file("imgs", "z.jpg", package = "gcnn")
+img <- torchvision::base_loader(img_path) |> torchvision::transform_to_tensor()
+img$permute(c(2, 3, 1)) |> as.array() |> as.raster() |> plot()
+```
+
+```{r, echo=FALSE, eval=TRUE, fig.alt = "A goat sitting comfortably on a meadow."}
+knitr::include_graphics("images/z.jpg")
+```
+
+First, we call `C_4$left_action_on_R2()` to rotate the grid.
+
+```{r}
+# Grid shape is [2, 1024, 1024], for a 2d, 1024 x 1024 image.
+img_grid_R2 <- torch::torch_stack(torch::torch_meshgrid(
+    list(
+      torch::torch_linspace(-1, 1, dim(img)[2]),
+      torch::torch_linspace(-1, 1, dim(img)[3])
+    )
+))
+
+# Transform the image grid with the matrix representation of some group element.
+transformed_grid <- C_4$left_action_on_R2(C_4$inverse(g1)$unsqueeze(1), img_grid_R2)
+
+```
+
+Second, we re-sample the image on the transformed grid. The goat now looks up to the sky.
+
+```{r}
+transformed_img <- torch::nnf_grid_sample(
+  img$unsqueeze(1), transformed_grid,
+  align_corners = TRUE, mode = "bilinear", padding_mode = "zeros"
+)
+
+transformed_img[1,..]$permute(c(2, 3, 1)) |> as.array() |> as.raster() |> plot()
+```
+
+```{r, echo=FALSE, eval=TRUE, fig.alt = "Same goat, rotated up by 90 degrees."}
+knitr::include_graphics("images/z2.jpg")
+```
+
+## Step 2: The lifting convolution
+
+We want to make use of existing, efficient `torch` functionality as much as possible. Concretely, we want to use `nn_conv2d()`. What we need, though, is a convolution kernel that's equivariant not just to translation, but also, to the action of $C_4$. This can be achieved by having one kernel each for every possible rotation.
+
+Implement that idea is exactly what `LiftingConvolution` does. The principle is the same as before: First, the grid is rotated, and then, the kernel (weight matrix) is re-sampled to the transformed grid.
+
+Why, though, call this a *lifting convolution*? The usual convolution kernel operates on $\mathbb{R}^2$; our extended version operates on combinations of $\mathbb{R}^2$ and $C_4$. In math speak, it has been *lifted* to the [semi-direct product](https://en.wikipedia.org/wiki/Semidirect_product) $\mathbb{R}^2\rtimes C_4$.
+
+```{r}
+lifting_conv <- LiftingConvolution(
+    group = CyclicGroup(order = 4),
+    kernel_size = 5,
+    in_channels = 3,
+    out_channels = 8
+  )
+
+x <- torch::torch_randn(c(2, 3, 32, 32))
+y <- lifting_conv(x)
+y$shape
+```
+
+    [1]  2  8  4 28 28
+
+Since, internally, `LiftingConvolution` uses an additional dimension to realize the product of translations and rotations, the output is not four-, but five-dimensional.
```

Comment:
```suggestion
Since, internally, `LiftingConvolution()` uses an additional dimension to realize the product of translations and rotations, the output is not four-, but five-dimensional.
```

Reply:
Well, this is sort of a corner case, you know ... it's a function, but it creates a module, a neural-network layer ... which we tend to think about as "agents" ... E.g. in Python, it's a class.
So yeah, I considered both ways, but all in all, I think this one should read better ...